### PR TITLE
Use more permissive email regex in strategy, issue #46

### LINKED ISSUE #46: More permissive email regex
The [email regex](https://github.com/TechnologyAdvice/obey/blob/d8faf05c36d8948b945fbb7055324b791f186770/src/typeStrategies/email.js#L3) included with obey does not match many valid emails.  I've created a regexr with valid email addresses for example: http://regexr.com/3d9cg

This was copied to one of our FE apps for validation, though it is too restrictive to use for validation as it would prevent users from submitting valid email addresses.  We may use it as a warning, but I think we should consider a more permissive rule that does not rule out valid email addresses.  If we can get a rule that does not rule out _any_ valid email, then we can use it for submission validation.

### PR DESCRIPTION
**Pull request by Fluidbyte**
Closes #46

### CODE REVIEW

File: .snyk [hunk 1]
```diff
@@ -1,3 +1,7 @@
-ignore: {}
+ignore:
+  'npm:marked:20150520':
+    - 'jsdoc > marked':
+        reason: Dev Dependecy
+        expires: '2016-05-28T00:25:33.252Z'
```
Comment (jhabdas) [inline .snyk]: Looks like [Marked got marked](https://snyk.io/vuln/npm:marked:20150520). I'd never seen `synk` until now. But I was wondering how those badges were created!

**Comment by jhabdas** [comment]
🍶

**Comment by davezuko** [comment]
I still lean against such extremely permissive validation (in this case it might as well not exist IMO), but at least it clears up some of the more common edge cases.

**Comment by Fluidbyte** [comment]
Agreed, I would tend to believe that the chances of someone using `"#&$*%&!!"+foo@goat.chee.se` would be rare and should most likely signify a BS email anyways, but there is a lot of noise in area of email validation about being more permissive

**Comment by levithomason** [comment]
🉑 would rather no validation than reject 1 good email.

**Comment by davezuko** [comment]
I believe that becomes a business decision at that point, however.

**Comment by TomFrost** [comment]
There are some things that just purely suck to validate with a regex, and email is one of those... so I'd be in favor of a lightweight and ridiculously fast dependency, but in the event that no such thing is super readily available, regex works to an acceptable level.

I'm with Levi -- this shouldn't turn away anything that even might be fringe-valid, because we simply can't assume anything about other people's use cases, and this project is public. What we COULD do, though, is apply a little more restriction to the _right_ side of the @. Unless I'm mistaken, spaces (for example) are illegal in hostnames. There might be some other ones, but just flipping `.+` to `\S+` wherever it appears on the right of the @ would be valuable.

**Comment by TomFrost** [comment]
(To clarify the above, it's worth looking what what characters _are_ actually legal in domain names, as long as that's a quick hunt-- the more symbols we can disallow there, the better)

**Comment by levithomason** [comment]
@TomFrost We're still knowingly excluding valid emails, see the [current RegExr](http://regexr.com/3datb).  If you're wanting a truly correct email validator for all public use cases, you'd need to allow dotless domains.  I don't know why there is still doubt as to whether or not these are valid, the [specs](https://tools.ietf.org/html/rfc5321#section-2.3.5) are clear:

>...In the case of a top-level domain used by itself in an email address, **a single string is used without any dots.**...

There are also examples like [this guy](http://serverfault.com/questions/309612/how-do-you-register-a-domain-without-a-tld) puzzled by finding one, and [this guy](http://isemail.info/foo%40bar#comment-294596693) upset that his TLD email doesn't work in most places.  Only as of Aug 2013 were [dotless domains prohibited](https://www.icann.org/news/announcement-2013-08-30-en) for _new gTLDs_, however, older gTLDs and ccTLDs are not subject to the prohibition.

**http://isemail.info/foo@bar** gets it right, including noting that the email is unusual and citing the related specifications that permit it but make it unusual.  Additionally, someone has ported it to npm as [isemail](https://www.npmjs.com/package/isemail).  Seems like overkill @ >1k lines to me.

If you're wanting an email validator that doesn't exclude valid addresses, the current regex isn't it.

```
.+@(\S+[^.]|\S)$
```

This ^ allows dotless domains, excludes whitespace in the domain, and excludes ending with a dot.  This is not a validator, but it would no longer exclude any valid emails as seen in this [RegExr](http://regexr.com/3datt).

image

I'll leave it in your court.

**Comment by levithomason** [comment]
aaaaaaaand just like that, it fails, allowing this:

```
a@b.c.
```

**http://isemail.info/a@b.c.** gets it exactly right, again, specs included.  I'll be done playing with RegEx where I shouldn't be.  Case in point, don't try to regex email addresses.

**Comment by Fluidbyte** [comment]
Right, I think this all falls into "business decision" category. This is a library and should do some basic, low-level checks that people can use "out of the box". For our internal use this may not work, so we build a custom type strategy that either uses our own, tailored RegEx or (even better) uses a service like we have discussed.

**Pull request merged**